mirgecom/profiling.py

```diff
@@ -0,0 +1,175 @@
+            num_values = len(value)
+
+            times = [v['time'] for v in value]
+            flops = [v['flops'] for v in value]
+            bytes_accessed = [v['bytes_accessed'] for v in value]
+
+            tbl.add_row([key.name, num_values, min(times),
+                int(mean(times)), max(times), min(flops), int(mean(flops)),
+                max(flops), min(bytes_accessed), int(mean(bytes_accessed)),
+                max(bytes_accessed), round(mean(bytes_accessed)/mean(times), 3)])
+
+        if hasattr(pytools.Table, 'github_markdown'):
+            print(tbl.github_markdown())
+        else:
+            print(tbl)
+
+    def call_loopy(self, program, **kwargs) -> dict:
+        program = self.transform_loopy_program(program)
+        assert program.options.return_dict
+        assert program.options.no_numpy
+
+        args_tuple = tuple(
+            (key, value.shape) if hasattr(value, 'shape') else (key, value)
+            for key, value in kwargs.items())
+
+        # Check if we need to get the invoker code to generate integer arguments
+        # N.B.: The invoker code might be different for the same program with
+        # different kwargs
+        if (program not in self.kernel_stats
+                or args_tuple not in self.kernel_stats[program]):
+            executor = program.target.get_kernel_executor(program, self.queue)
+            info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+            kernel = executor.get_typed_and_scheduled_kernel(
+                executor.arg_to_dtype_set(kwargs))
+
+            data = info.implemented_data_info
+
+            # Generate the wrapper code
+            wrapper = executor.get_wrapper_generator()
+
+            gen = PythonCodeGenerator()
+
+            wrapper.generate_integer_arg_finding_from_shapes(gen, kernel, data)
+            wrapper.generate_integer_arg_finding_from_offsets(gen, kernel, data)
+            wrapper.generate_integer_arg_finding_from_strides(gen, kernel, data)
+
+            types = {}
+            param_dict = {}
+
+            for key, value in kwargs.items():
+                if hasattr(value, 'dtype') and not value.dtype == object:
+                    types[key] = value.dtype
+                param_dict[key] = value
+
+            for key, value in kernel.arg_dict.items():
+                if key not in param_dict:
+                    param_dict[key] = None
+
+            for d in data:
+                if d.name not in param_dict:
+                    param_dict[d.name] = None
+
+            # Run the wrapper code, save argument values in param_dict
+            exec(gen.get(), param_dict)
+
+            # Get flops/memory statistics
+            kernel = lp.add_and_infer_dtypes(kernel, types)
+            op_map = lp.get_op_map(kernel, subgroup_size='guess')
+            bytes_accessed = lp.get_mem_access_map(kernel, subgroup_size='guess') \
+              .to_bytes().eval_and_sum(param_dict)
```

Comment:
We should also gather the size of the accessed footprint.

Reply:
Running

```python
footprint_bytes = lp.gather_access_footprint_bytes(kernel, True)
```

at this point gives the following error:

```
/Users/mdiener/Work/emirge/loo-py/loopy/statistics.py:1217: LoopyWarning: in kernel resample_by_picking: Barvinok wrappers are not installed. Counting routines have overestimated the number of integer points in your loop domain. (add 'count_overestimate' to silenced_warnings kernel attribute to disable)
  warn_with_kernel(kernel, "count_overestimate",
Traceback (most recent call last):
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 126, in __call__
    method = getattr(self, expr.mapper_method)
AttributeError: 'tuple' object has no attribute 'mapper_method'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mdiener/Work/emirge/loo-py/loopy/symbolic.py", line 1667, in get_access_range
    idx_aff = guarded_aff_from_expr(access_map.space, subscript[idim])
  File "/Users/mdiener/Work/emirge/loo-py/loopy/symbolic.py", line 1429, in guarded_aff_from_expr
    return with_aff_conversion_guard(aff_from_expr, space, expr, vars_to_zero)
  File "/Users/mdiener/Work/emirge/loo-py/loopy/symbolic.py", line 1420, in with_aff_conversion_guard
    raise ExpressionToAffineConversionError(
loopy.diagnostic.ExpressionToAffineConversionError: could not convert expression 'to_element_indices[iel]' to affine representation: UnknownVariableError: to_element_indices

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "examples/wave-eager.py", line 124, in <module>
    main(True)
  File "examples/wave-eager.py", line 107, in main
    fields = rk4_step(fields, t, dt, rhs)
  File "/Users/mdiener/Work/emirge/mirgecom/mirgecom/integrators.py", line 40, in rk4_step
    k1 = rhs(t, state)
  File "examples/wave-eager.py", line 101, in rhs
    return wave_operator(discr, c=1, w=w)
  File "/Users/mdiener/Work/emirge/mirgecom/mirgecom/wave.py", line 79, in wave_operator
    dir_u = discr.project("vol", BTAG_ALL, u)
  File "/Users/mdiener/Work/emirge/grudge/grudge/eager.py", line 73, in project
    return self.connection_from_dds(src, tgt)(vec)
  File "/Users/mdiener/Work/emirge/meshmode/meshmode/discretization/connection/direct.py", line 403, in __call__
    actx.call_loopy(pick_knl(),
  File "/Users/mdiener/Work/emirge/mirgecom/mirgecom/profiling.py", line 160, in call_loopy
    footprint_bytes = lp.gather_access_footprint_bytes(kernel, True)
  File "/Users/mdiener/Work/emirge/loo-py/loopy/statistics.py", line 1851, in gather_access_footprint_bytes
    fp = gather_access_footprints(kernel,
  File "/Users/mdiener/Work/emirge/loo-py/loopy/statistics.py", line 1815, in gather_access_footprints
    write_footprints.append(afg(insn.assignees))
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 132, in __call__
    return self.map_foreign(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 182, in map_foreign
    return self.map_tuple(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 297, in map_list
    return self.combine(self.rec(child, *args, **kwargs) for child in expr)
  File "/Users/mdiener/Work/emirge/loo-py/loopy/statistics.py", line 1090, in combine
    return reduce(merge_dicts, values)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 297, in <genexpr>
    return self.combine(self.rec(child, *args, **kwargs) for child in expr)
  File "/Users/mdiener/Work/emirge/pymbolic/pymbolic/mapper/__init__.py", line 134, in __call__
    return method(expr, *args, **kwargs)
  File "/Users/mdiener/Work/emirge/loo-py/loopy/statistics.py", line 1107, in map_subscript
    access_range = get_access_range(self.domain, subscript,
  File "/Users/mdiener/Work/emirge/loo-py/loopy/symbolic.py", line 1679, in get_access_range
    raise UnableToDetermineAccessRange(
loopy.symbolic.UnableToDetermineAccessRange: unable to determine access range of subscript: [to_element_indices[iel], idof_inner + idof_outer*16] (encountered ExpressionToAffineConversionError: could not convert expression 'to_element_indices[iel]' to affine representation: UnknownVariableError: to_element_indices)
```

Reply:
That's expected (as in, catch that exception and record no value). Projection operators by necessity include indirect (i.e. data-dependent) addressing. In that situation (and without further information), it's simply not possible to statically determine the access footprint.

Reply:
OIC. I thought that `ignore_uncountable` would already take care of that.

Reply:
Added.